app/src/main/java/xyz/block/gosling/features/agent/Tools.kt

```diff
@@ -81,83 +79,91 @@ object ToolHandler {
         return gestureResult
     }
 
-    private fun serializeNodeHierarchy(node: AccessibilityNodeInfo): NodeInfo {
-        try {
-            val bounds = Rect().also { node.getBoundsInScreen(it) }
-
-            return NodeInfo(
-                className = node.className?.toString(),
-                packageName = node.packageName?.toString(),
-                text = node.text?.toString()?.takeIf { it.isNotEmpty() },
-                contentDesc = node.contentDescription?.toString()?.takeIf { it.isNotEmpty() },
-                resourceId = node.viewIdResourceName?.takeIf { it.isNotEmpty() },
-                bounds = NodeBounds(
-                    left = bounds.left,
-                    top = bounds.top,
-                    right = bounds.right,
-                    bottom = bounds.bottom
-                ),
-                clickable = if (node.isClickable) true else null,
-                focusable = if (node.isFocusable) true else null,
-                scrollable = if (node.isScrollable) true else null,
-                editable = if (node.isEditable) true else null,
-                enabled = if (!node.isEnabled) false else null,
-                children = if (node.childCount > 0) {
-                    (0 until node.childCount).mapNotNull { i ->
-                        node.getChild(i)?.let { childNode ->
-                            try {
-                                serializeNodeHierarchy(childNode)
-                            } catch (e: Exception) {
-                                NodeInfo(
-                                    error = "Failed to serialize child node: ${e.message}",
-                                    bounds = NodeBounds(0, 0, 0, 0)
-                                )
-                            }
-                        }
-                    }
-                } else null
-            )
-        } catch (e: Exception) {
-            return NodeInfo(
-                error = "Failed to serialize node: ${e.message}",
-                bounds = NodeBounds(0, 0, 0, 0)
-            )
-        }
-    }
-
     @Tool(
         name = "getUiHierarchy",
-        description = "Get the current UI hierarchy as a JSON string. " +
-                "This shows all UI elements, their properties, and locations on screen. " +
-                "clickable, focusable, scrollable, and editable properties are only mentioned " +
-                "when true, enabled only when false.",
+        description = "Get the current UI hierarchy in a compact format. " +
+                "Shows UI elements with their properties and locations on screen in a hierarchical structure. " +
+                "Only includes relevant properties (clickable, focusable, etc. when true; enabled when false).",
```

Comment:
It probably doesn't matter all that much, but we could spend some more time on the prompt here. The AI doesn't care it is compact, it doesn't know it wasn't before; same for relevant. To really find out if we can improve we'd need some failure cases but for now, I would expand this a bit and describe for example that the bounding box is [left,top,right,bottom] and not [left,top,width,height] for example

Reply:
yes - I agree - and we can compress the exact format of the data with short hand notation I expect if we explain it here.

Reply:
agreed. the other thing to think about, is to provide it with just a click target. I looked at the log and if the rectange is [left, top, bottom, right], it will click on [left, top]. And sometimes it misses because there is something over it. So if we would for every clickable element find the mid point of the visible area, we might be doing a slightly better job.

Reply:
adding midpoint calculated